File: lib/aura/kernel/job.rb
# frozen_string_literal: true

require "securerandom"

module Aura
  module Kernel
    class Job
      attr_reader :id, :status, :started_at, :ended_at, :events, :metadata

      def initialize(metadata = {})
        @id = SecureRandom.uuid
        @status = :pending # :pending, :running, :completed, :failed
        @started_at = nil
        @ended_at = nil
        @events = []
        @metadata = metadata
      end

      def start!
        @started_at = Time.now
        @status = :running
      end

      def complete!
        @ended_at = Time.now
        @status = :completed
      end

      def fail!(error)
        @ended_at = Time.now
        @status = :failed
        @metadata[:error] = error.message
      end

      def add_event(event_id)
        @events << event_id
      end

      def to_h
        {
          id: @id,
          status: @status,
          started_at: @started_at,
          ended_at: @ended_at,
          events: @events,
          metadata: @metadata
        }
      end
    end
  end
end

File: src/core/kernel/job.ts
import crypto from 'crypto';

export type JobStatus = 'pending' | 'running' | 'completed' | 'failed';

export class Job {
  public readonly id: string;
  public status: JobStatus;
  public started_at: Date | null = null;
  public ended_at: Date | null = null;
  public readonly events: number[] = [];
  public metadata: Record<string, any>;

  constructor(metadata: Record<string, any> = {}) {
    this.id = crypto.randomUUID();
    this.status = 'pending';
    this.metadata = metadata;
  }

  public start(): void {
    this.started_at = new Date();
    this.status = 'running';
  }

  public complete(): void {
    this.ended_at = new Date();
    this.status = 'completed';
  }

  public fail(error: Error | string): void {
    this.ended_at = new Date();
    this.status = 'failed';
    this.metadata.error = typeof error === 'string' ? error : error.message;
  }

  public addEvent(eventId: number): void {
    this.events.push(eventId);
  }

  public toObject(): any {
    return {
      id: this.id,
      status: this.status,
      started_at: this.started_at,
      ended_at: this.ended_at,
      events: this.events,
      metadata: this.metadata,
    };
  }
}
